### ISSUE TITLE
Some regression tests fail (locally) on current master

### ISSUE THREAD

**Issue by ansgar-t** [explosion/spaCy#2379]
## How to reproduce the behaviour
Here's what I did: 
```
git clone https://github.com/explosion/spaCy.git explosion.spaCy
mkvirtualenv -p /usr/local/opt/python3/bin/python3 spaCyTmp 
cd explosion.spaCy/
pip install -r requirements.txt
python setup.py build_ext --inplace
python -m spacy download en_core_web_md
python -m spacy download en_core_web_sm
python -m spacy download fr_core_news_sm
python -m spacy download es_core_news_sm
python -m spacy download es_core_news_md
python -m spacy download en
python -m spacy download de
python -m spacy download fr
python -m spacy download es
py.test spacy --models --all --slow
```

This results in some test failures:
```
====================================================================================================================== FAILURES ======================================================================================================================
________________________________________________________________________________________________ test_issue401[en_core_web_sm-Jane's got a new car-1] ________________________________________________________________________________________________

EN = <spacy.lang.en.English object at 0x11da6b518>, text = "Jane's got a new car", i = 1

    @pytest.mark.models('en')
    @pytest.mark.parametrize('text,i', [("Jane's got a new car", 1),
                                        ("Jane thinks that's a nice car", 3)])
    def test_issue401(EN, text, i):
        """Text that 's in contractions is not lemmatized as '."""
        tokens = EN(text)
>       assert tokens[i].lemma_ != "'"
E       AttributeError: 'str' object has no attribute 'lemma_'

spacy/tests/regression/test_issue401.py:13: AttributeError
___________________________________________________________________________________________ test_issue401[en_core_web_sm-Jane thinks that's a nice car-3] ____________________________________________________________________________________________

EN = <spacy.lang.en.English object at 0x11da6b518>, text = "Jane thinks that's a nice car", i = 3

    @pytest.mark.models('en')
    @pytest.mark.parametrize('text,i', [("Jane's got a new car", 1),
                                        ("Jane thinks that's a nice car", 3)])
    def test_issue401(EN, text, i):
        """Text that 's in contractions is not lemmatized as '."""
        tokens = EN(text)
>       assert tokens[i].lemma_ != "'"
E       AttributeError: 'str' object has no attribute 'lemma_'

spacy/tests/regression/test_issue401.py:13: AttributeError
____________________________________________________________________________________________________ test_issue686[en_core_web_sm-He is the man] _____________________________________________________________________________________________________

EN = <spacy.lang.en.English object at 0x11da6b518>, text = 'He is the man'

    @pytest.mark.models('en')
    @pytest.mark.parametrize('text', ["He is the man", "he is the man"])
    def test_issue686(EN, text):
        """Test that pronoun lemmas are assigned correctly."""
        tokens = EN(text)
>       assert tokens[0].lemma_ == "-PRON-"
E       AttributeError: 'str' object has no attribute 'lemma_'

spacy/tests/regression/test_issue686.py:12: AttributeError
____________________________________________________________________________________________________ test_issue686[en_core_web_sm-he is the man] _____________________________________________________________________________________________________

EN = <spacy.lang.en.English object at 0x11da6b518>, text = 'he is the man'

    @pytest.mark.models('en')
    @pytest.mark.parametrize('text', ["He is the man", "he is the man"])
    def test_issue686(EN, text):
        """Test that pronoun lemmas are assigned correctly."""
        tokens = EN(text)
>       assert tokens[0].lemma_ == "-PRON-"
E       AttributeError: 'str' object has no attribute 'lemma_'

spacy/tests/regression/test_issue686.py:12: AttributeError
______________________________________________________________________________________________ test_issue717[en_core_web_sm-You're happy-You are happy] ______________________________________________________________________________________________

EN = <spacy.lang.en.English object at 0x11da6b518>, text1 = "You're happy", text2 = 'You are happy'

    @pytest.mark.models('en')
    @pytest.mark.parametrize('text1,text2',
        [("You're happy", "You are happy"),
         ("I'm happy", "I am happy"),
         ("he's happy", "he's happy")])
    def test_issue717(EN, text1, text2):
        """Test that contractions are assigned the correct lemma."""
        doc1 = EN(text1)
        doc2 = EN(text2)
>       assert doc1[1].lemma_ == doc2[1].lemma_
E       AttributeError: 'str' object has no attribute 'lemma_'

spacy/tests/regression/test_issue717.py:16: AttributeError
_________________________________________________________________________________________________ test_issue717[en_core_web_sm-I'm happy-I am happy] _________________________________________________________________________________________________

EN = <spacy.lang.en.English object at 0x11da6b518>, text1 = "I'm happy", text2 = 'I am happy'

    @pytest.mark.models('en')
    @pytest.mark.parametrize('text1,text2',
        [("You're happy", "You are happy"),
         ("I'm happy", "I am happy"),
         ("he's happy", "he's happy")])
    def test_issue717(EN, text1, text2):
        """Test that contractions are assigned the correct lemma."""
        doc1 = EN(text1)
        doc2 = EN(text2)
>       assert doc1[1].lemma_ == doc2[1].lemma_
E       AttributeError: 'str' object has no attribute 'lemma_'

spacy/tests/regression/test_issue717.py:16: AttributeError
________________________________________________________________________________________________ test_issue717[en_core_web_sm-he's happy-he's happy] _________________________________________________________________________________________________

EN = <spacy.lang.en.English object at 0x11da6b518>, text1 = "he's happy", text2 = "he's happy"

    @pytest.mark.models('en')
    @pytest.mark.parametrize('text1,text2',
        [("You're happy", "You are happy"),
         ("I'm happy", "I am happy"),
         ("he's happy", "he's happy")])
    def test_issue717(EN, text1, text2):
        """Test that contractions are assigned the correct lemma."""
        doc1 = EN(text1)
        doc2 = EN(text2)
>       assert doc1[1].lemma_ == doc2[1].lemma_
E       AttributeError: 'str' object has no attribute 'lemma_'

spacy/tests/regression/test_issue717.py:16: AttributeError
_________________________________________________________________________________________________________ test_issue719[en_core_web_sm-s...] _________________________________________________________________________________________________________

EN = <spacy.lang.en.English object at 0x11da6b518>, text = 's...'

    @pytest.mark.models('en')
    @pytest.mark.parametrize('text', ["s..."])
    def test_issue719(EN, text):
        """Test that the token 's' is not lemmatized into empty string."""
        tokens = EN(text)
>       assert tokens[0].lemma_ != ''
E       IndexError: string index out of range

spacy/tests/regression/test_issue719.py:12: IndexError
___________________________________________________________________________________________________________ test_issue955[en_core_web_sm] ____________________________________________________________________________________________________________

EN = <spacy.lang.en.English object at 0x11da6b518>

    @pytest.mark.models('en')
    def test_issue955(EN):
        '''Test that we don't have any nested noun chunks'''
        doc = EN('Does flight number three fifty-four require a connecting flight'
                 ' to get to Boston?')
        seen_tokens = set()
>       for np in doc.noun_chunks:
E       AttributeError: 'str' object has no attribute 'noun_chunks'

spacy/tests/regression/test_issue995.py:12: AttributeError
==================================================================================== 9 failed, 1408 passed, 132 skipped, 46 xfailed, 10 xpassed in 298.42 seconds ====================================================================================
```

## Info about spaCy

* **Python version:** 3.5.1
* **spaCy version:** 2.0.12.dev0
* **Models:** de, en, en_core_web_md, en_core_web_sm, es, es_core_news_md, es_core_news_sm, fr, fr_core_news_sm
* **Platform:** Darwin-16.7.0-x86_64-i386-64bit

**Comment by ansgar-t** [comment]
However, running the first failing test file in isolation results in 2 passes: 
```
py.test spacy/tests/regression/test_issue401.py --models --all --slow
```

**Comment by ines** [comment]
Thanks! I *think* this might be related to some specifics of pytest and the way fixtures are handled, reused and/or initialised? Now that I think about it, I vaguely remember coming across a similar issue and being puzzled by it.

We'll definitely spend some more time trying to get to the bottom of this. The model tests need an overhaul anyway, so this is good timing. I'll also tag this issue `help wanted` in case someone from the community has an idea what might be going on here.

**Comment by mattmcgrattan** [comment]
I'm having a similar problem. I have a couple of modules of tests in which I create a Spacy instance, add some pipeline steps, and then test against them to check the output of each pipeline step.

If I run the files independently, one at a time, all tests pass. If I run them together, some of them fail. In exactly the way that @ansgar-t describes in https://github.com/explosion/spaCy/issues/2379#issuecomment-392322391

These are my tests, not Spacy's, but the same issue occurs. I've tried running the tests with unittest, with py.test, and with Nose.

**Comment by ines** [comment]
@mattmcgrattan Thanks for sharing – do you happen to have a more or less reproducible setup that consistently fails? And are you also using `pytest` fixtures for the models you're loading?

One conclusion I've come to so far is that for some reason, the model fixtures like `EN` must return an empty string. This is the only possible value for an error like this where index `0` is out of range, right?

```
assert tokens[0].lemma_ != ''
IndexError: string index out of range
```

**Comment by mattmcgrattan** [comment]
I'll see if I can create a reproducible version of my issue. The pipeline components I'm using haven't been open sourced yet, but I can probably make a cut down version to reproduce the test error.

**Comment by ines** [comment]
Thanks a lot! 🙏 And I'm pretty sure it's not related to the pipeline components, just the loaded model / the `nlp` object with a loaded model.

I'll also try to read up on some of the inner workings of `pytest` and if there's something that could potentially explain what's going on here...

**Comment by mattmcgrattan** [comment]
@ines I think I've identified the root cause of my error, and it's looking like it's an issue with mutable defaults on Span extensions. Which is not a bug with Spacy (obviously). I need to review my code, as I am using getters and setters, but I think that's the most likely explanation, that somewhere there's an extension being set on the global Span object that's polluting the results between concurrent/consecutive tests.

**Comment by ines** [comment]
@mattmcgrattan Thanks for the update – that's interesting! I still haven't fully figured out what the problem was in our case, but it hasn't come up again since the test refactoring.

I just published our new model test suite in the `spacy-models` repo:
https://github.com/explosion/spacy-models/tree/master/tests

We refactored all tests and updated the pytest framework around them to make it easier to execute specific model tests. We're already using the tests internally as part of our model training pipeline for [v2.1.0](https://github.com/explosion/spaCy/releases/tag/v2.1.0a0) and they've been super helpful so far. We're also hoping that the new setup will make it easier for users to test their custom trained models. (For example, there are a bunch of general-purpose "sanity checks" – if your model doesn't pass these, you know it's broken).

My PR in #2568 removes the model tests from the main repo and also tidies up the existing test suite to make it faster and more reliable.
